### PR TITLE
Reuse program structure

### PR DESCRIPTION
**Pull request by vladima**
Currently on every edit `createProgram` revisits all source files in order to discover all dependencies that are absent in root files and reconstruct the proper sequence of files. This is right thing to do however usually most of edits don't modify imports and tripleslash references so set and order of files remains the same. This PR tries to detect such cases: if edit does not affect the structure of the program then we'll just reuse nonchanged files from the old program keeping the same order.

### CODE REVIEW

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -5634,7 +5634,6 @@ namespace ts {
             // will immediately bail out of walking any subtrees when we can see that their parents
             // are already correct.
             let result = Parser.parseSourceFile(sourceFile.fileName, newText, sourceFile.languageVersion, syntaxCursor, /* setParentNode */ true)
-
```
Comment (vladima) [inline src/compiler/parser.ts]: unnecessary change, will remove it from the PR

File: src/compiler/program.ts [hunk 2]
```diff
@@ -160,15 +162,22 @@ namespace ts {
...
-        // Do not process the default library if:
-        //  - The '--noLib' flag is used.
-        //  - A 'no-default-lib' reference comment is encountered in
-        //      processing the root files.
-        if (!skipDefaultLib) {
-            processRootFile(host.getDefaultLibFileName(options), /*isDefaultLib:*/ true);
+        
+        // if old program was provided by has different target module kind - assume that it cannot be reused
+        // different module kind can lead to different way of resolving modules
+        if (oldProgram && oldProgram.getCompilerOptions().module !== options.module) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: why not move this to tryReuseStructureFromOldProgram and just return false if this true.

Comment (vladima) [inline src/compiler/program.ts]: because in case if structure cannot be reused we still want to utilize previous resolutions, see `processImports`. However if old version of the program targeted different module kind - then resolutions should not be shared (not true currently but will be true after we incorporate different module resolution strategies)

File: src/compiler/program.ts [hunk 3]
```diff
@@ -211,6 +221,77 @@ namespace ts {
...
+                    // different order of root names - structure cannot be reused
+                    return false;
+                }
+            }
+            
+            // check if program source files has changed in the way that can affect structure of the program
+            let newSourceFiles: SourceFile[] = [];
+            for (let oldSourceFile of oldProgram.getSourceFiles()) {
+                let newSourceFile: SourceFile;
+                if (host.hasChanges(oldSourceFile)) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: how about doing referential comparison to know if anything changed instead of asking the host? e.g.

```ts
newSourceFile = host.getSourceFile(oldSourceFile.fileName, options.target);
if (newSourceFile !== oldSourceFile) {
    if (!newSourceFile) {
         return false;
    }
     // check tripleslash references
    ....
     // check imports
     .....
}
else {
    // file has no changes - use it as is
    newSourceFile = oldSourceFile;
}
```

Comment (vladima) [inline src/compiler/program.ts]: fair, will do

File: src/compiler/program.ts [hunk 4]
```diff
@@ -211,6 +221,77 @@ namespace ts {
...
+            }
+            
+            // update fileName -> file mapping
+            for (let file of newSourceFiles) {
+                filesByName.set(file.fileName, file);
+            }
+            
+            files = newSourceFiles;
+            
+            oldProgram.structureIsReused = true;
```
Comment (mhegazy) [inline src/compiler/program.ts]: what does this give us at this point? we have copied the data, there should not be any dependency, correct?

File: src/compiler/program.ts [hunk 5]
```diff
@@ -332,6 +413,62 @@ namespace ts {
...
         function processRootFile(fileName: string, isDefaultLib: boolean) {
             processSourceFile(normalizePath(fileName), isDefaultLib);
+        }        
+    
+        function fileReferenceIsEqualTo(a: FileReference, b: FileReference): boolean {
+            return a.fileName === b.fileName;
+        }
+        
+        function moduleNameIsEqualTo(a: LiteralExpression, b: LiteralExpression): boolean {
+            return a.text ===b.text;
```
Comment (mhegazy) [inline src/compiler/program.ts]: nit: space after `===`

File: src/compiler/program.ts (hunk 4 above)
Comment (vladima) [inline src/compiler/program.ts]: this flag is currently used only in tests

File: src/compiler/program.ts [hunk 6]
```diff
@@ -211,6 +221,77 @@ namespace ts {
...
+
+            // there is an old program, check if we can reuse its structure
+            let oldRootNames = oldProgram.getRootFileNames();
+            if (rootNames.length !== oldRootNames.length) {
+                // different amount of root names - structure cannot be reused
+                return false;
+            }
+            
+            for (let i = 0; i < rootNames.length; i++) {
+                if (oldRootNames[i] !== rootNames[i]) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: why not just use `arrayIsEqualTo` here as well?

File: src/compiler/program.ts (hunk 3 above)
Comment (vladima) [inline src/compiler/program.ts]: technically there is difference, in LS scenarios querying for version can be slightly less expensive because we this is just a lookup in host cache where is getting new file involves doing more stuff

File: src/compiler/program.ts (hunk 6 above)
Comment (vladima) [inline src/compiler/program.ts]: will do

File: src/compiler/program.ts [hunk 7]
```diff
@@ -332,6 +413,62 @@ namespace ts {
...
+                            break;
+                        }
+                        if (!(<LiteralExpression>moduleNameExpr).text) {
+                            break;
+                        }
+
+                        (imports || (imports = [])).push(<LiteralExpression>moduleNameExpr);
+                        break;
+                    case SyntaxKind.ModuleDeclaration:
+                        if ((<ModuleDeclaration>node).name.kind === SyntaxKind.StringLiteral && (node.flags & NodeFlags.Ambient || isDeclarationFile(file))) {
```
Comment (mhegazy) [inline src/compiler/program.ts]: I know that this part has been around already before your change, but i think we do not need the fork. we just need a helper, given a node find all the external module references inside, if we find an external module declaration recursively call the same routine on the module body.

File: src/compiler/program.ts (hunk 5 above)
Comment (vladima) [inline src/compiler/program.ts]: done

File: src/compiler/utilities.ts [hunk 8]
```diff
@@ -78,6 +78,40 @@ namespace ts {
...
+        }
+
+        return true;
+    }    
+   
+    export function hasResolvedModuleName(sourceFile: SourceFile, moduleName: LiteralExpression): boolean {
+        return sourceFile.resolvedModules && hasProperty(sourceFile.resolvedModules, moduleName.text);
+    }
+
+    export function getResolvedModuleFileName(sourceFile: SourceFile, moduleName: LiteralExpression): string {
```
Comment (mhegazy) [inline src/compiler/utilities.ts]: nit: can we make this take moduleNameText instead of a node, this way it is easier to understand the logic looking at the call.

File: src/compiler/utilities.ts [hunk 9]
```diff
@@ -78,6 +78,40 @@ namespace ts {
...
+        return sourceFile.resolvedModules && sourceFile.resolvedModules[moduleName.text];
```
Comment (mhegazy) [inline src/compiler/utilities.ts]: should not this be hasOwnProperty to avoid modulename looks for something like "toString".

Comment (vladima) [inline src/compiler/utilities.ts]: done

File: src/compiler/utilities.ts (hunk 8 above)
Comment (vladima) [inline src/compiler/utilities.ts]: ok

File: src/compiler/utilities.ts (hunk 9 above)
Comment (mhegazy) [inline src/compiler/utilities.ts]: never mind. noticed that we are using `undefined` to mark missing.

File: src/compiler/program.ts (hunk 2 above)
Comment (mhegazy) [inline src/compiler/program.ts]: also noResolve can not change between the two.

Comment (mhegazy) [inline src/compiler/program.ts]: if you want to be conservative and check the whole compilerOptions properties i am fine with that.

File: src/compiler/program.ts (hunk 3 above)
Comment (mhegazy) [inline src/compiler/program.ts]: but the host (the ls in this case) can use the version to decide which reference to return. if the version is different, it is going to need incrementally parse anyways.

File: src/compiler/program.ts [hunk 10]
```diff
@@ -450,57 +587,83 @@ namespace ts {
...
-                                if (parentPath === searchPath) {
-                                    break;
+        
+        function processImportedModules(file: SourceFile, basePath: string) {            
+            collectExternalModuleReferences(file);            
+            if (file.imports.length) {               
+                let allImportsInTheCache = false;
+
+                // try to grab existing module resolutions from the old source file
+                let oldSourceFile: SourceFile = oldProgram && oldProgram.getSourceFile(file.fileName);
```
Comment (mhegazy) [inline src/compiler/program.ts]: why do we need to this here. can not we do this in tryReuseStructureFromOldProgram, just after we got the newSourceFile, we know the old file resolutions, we can do that there, so we have two helpers, processImports, and processImportsIncrementally.. this way the oldProgram, oldSoruceFile knowledge is kept in one place.

File: src/compiler/program.ts (hunk 3 above)
Comment (vladima) [inline src/compiler/program.ts]: sure, that is why I've said 'slightly' :)

**Comment by mhegazy** [comment]
We should also put this in tsc.ts in the `--watch` implementation. it should simplify it a bit.

File: src/compiler/program.ts (hunk 10 above)
Comment (vladima) [inline src/compiler/program.ts]: `tryReuseStructureFromOldProgram` can bail out before we get to this file

File: src/compiler/program.ts (hunk 2 above)
Comment (mhegazy) [inline src/compiler/program.ts]: and target.

File: src/services/services.ts [hunk 11]
```diff
@@ -2471,6 +2472,8 @@ namespace ts {
             let newSettings = hostCache.compilationSettings();
             let changesInCompilationSettingsAffectSyntax = oldSettings && oldSettings.target !== newSettings.target;
 
+            let reusableOldProgram = changesInCompilationSettingsAffectSyntax ? undefined : program;
```
Comment (mhegazy) [inline src/services/services.ts]: we can move the target check to create program right?

Comment (vladima) [inline src/services/services.ts]: sure

File: src/compiler/program.ts (hunk 7 above)
Comment (vladima) [inline src/compiler/program.ts]: I'd prefer to make it a separte PR

**Comment by mhegazy** [comment]
:+1:

File: src/compiler/program.ts [hunk 12]
```diff
@@ -160,15 +162,26 @@ namespace ts {
...
-        forEach(rootNames, name => processRootFile(name, /*isDefaultLib:*/ false));
-
-        // Do not process the default library if:
-        //  - The '--noLib' flag is used.
-        //  - A 'no-default-lib' reference comment is encountered in
-        //      processing the root files.
-        if (!skipDefaultLib) {
-            processRootFile(host.getDefaultLibFileName(options), /*isDefaultLib:*/ true);
+        
+        if (oldProgram) {
```
Comment (CyrusNajmabadi) [inline src/compiler/program.ts]: Just add an explanatory comment.  Something to give hte intuition for why you are checking these specific options, and the types of options you'd put here in the future.

File: src/compiler/program.ts [hunk 13]
```diff
@@ -211,6 +225,64 @@ namespace ts {
...
+
+                if (oldSourceFile !== newSourceFile) {
+                    // check tripleslash references
+                    if (!arrayIsEqualTo(oldSourceFile.referencedFiles, newSourceFile.referencedFiles, fileReferenceIsEqualTo)) {
+                        // tripleslash references has changed
+                        return false;
+                    }
+                    
+                    // check imports
+                    collectExternalModuleReferences(newSourceFile);
```
Comment (CyrusNajmabadi) [inline src/compiler/program.ts]: this step seems odd.  Are the imports not automatically collected?

File: src/compiler/program.ts [hunk 14]
```diff
@@ -211,6 +225,64 @@ namespace ts {
...
+                    }
+                    
+                    // check imports
+                    collectExternalModuleReferences(newSourceFile);                    
+                    if (!arrayIsEqualTo(oldSourceFile.imports, newSourceFile.imports, moduleNameIsEqualTo)) {
+                        // imports has changed
+                        return false;
+                    }
+                    // pass the cache of module resolutions from the old source file
+                    newSourceFile.resolvedModules = oldSourceFile.resolvedModules;
```
Comment (CyrusNajmabadi) [inline src/compiler/program.ts]: I think you may have to check for no-default-lib.

File: src/compiler/types.ts [hunk 15]
```diff
@@ -1235,6 +1244,7 @@ namespace ts {
         /* @internal */ getIdentifierCount(): number;
         /* @internal */ getSymbolCount(): number;
         /* @internal */ getTypeCount(): number;
+        /* @internal */ structureIsReused?: boolean;
```
Comment (CyrusNajmabadi) [inline src/compiler/types.ts]: Comment that this is for tests ony.

File: src/compiler/utilities.ts [hunk 16]
```diff
@@ -78,6 +78,41 @@ namespace ts {
...
+
+        return true;
+    }    
+   
+    export function hasResolvedModuleName(sourceFile: SourceFile, moduleNameText: string): boolean {
+        return sourceFile.resolvedModules && hasProperty(sourceFile.resolvedModules, moduleNameText);
+    }
+
+    export function getResolvedModuleFileName(sourceFile: SourceFile, moduleNameText: string): string {
+        return hasResolvedModuleName(sourceFile, moduleNameText) ? sourceFile.resolvedModules[moduleNameText]: undefined;
```
Comment (CyrusNajmabadi) [inline src/compiler/utilities.ts]: space or newline between ] and :

File: tests/cases/unittests/reuseProgramStructure.ts [hunk 17]
```diff
@@ -0,0 +1,261 @@
+/// <reference path="..\..\..\src\harness\external\mocha.d.ts" />
+/// <reference path='..\..\..\src\harness\harness.ts' />
+/// <reference path="..\..\..\src\harness\harnessLanguageService.ts" />
+
+module ts {
+	
+	const enum ChangedPart {
```
Comment (CyrusNajmabadi) [inline tests/cases/unittests/reuseProgramStructure.ts]: tabs!

File: src/compiler/program.ts (hunk 12 above)
Comment (vladima) [inline src/compiler/program.ts]: done

File: src/compiler/program.ts (hunk 14 above)
Comment (vladima) [inline src/compiler/program.ts]: done

File: src/compiler/types.ts (hunk 15 above)
Comment (vladima) [inline src/compiler/types.ts]: Done

File: src/compiler/utilities.ts (hunk 16 above)
Comment (vladima) [inline src/compiler/utilities.ts]: done

File: tests/cases/unittests/reuseProgramStructure.ts (hunk 17 above)
Comment (vladima) [inline tests/cases/unittests/reuseProgramStructure.ts]: oops...

**Comment by mhegazy** [comment]
this change moved to #4154

**Comment by mhegazy** [comment]
@vladima can we close this now

**Comment by vladima** [comment]
subsumed by #4154
